sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1211,15 +1211,23 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
         """
   }
 
-  private[this] def castToIntervalCode(from: DataType): CastFunction = from match {
+  private[this] def castToIntervalCode(
+      from: DataType,
+      ctx: CodegenContext): CastFunction = from match {
     case StringType =>
       val util = IntervalUtils.getClass.getCanonicalName.stripSuffix("$")
+      val calOpt = ctx.freshVariable("calOpt", classOf[Option[CalendarInterval]])
+
       (c, evPrim, evNull) =>
-        code"""$evPrim = $util.safeFromString($c.toString());
-           if(${evPrim} == null) {
-             ${evNull} = true;
-           }
-         """.stripMargin
+        code"""
+          scala.Option<org.apache.spark.unsafe.types.CalendarInterval> $calOpt =
```

Comment:
if it's used in codegen, I think it's more natural to return null instead of None.